pilot/pkg/proxy/envoy/proxy.go

```diff
@@ -140,9 +143,18 @@ func (e *envoy) Run(config interface{}, epoch int, abort <-chan error) error {
 	if err := cmd.Start(); err != nil {
 		return err
 	}
+	// record envoy process number
+	e.processNum.Inc()
 	done := make(chan error, 1)
 	go func() {
-		done <- cmd.Wait()
+		err := cmd.Wait()
+		e.processNum.Dec()
+		if err == nil {
+			if e.processNum.Load() == 0 {
+				err = fmt.Errorf("envoy process is killed, maybe by SIGTERM")
+			}
+		}
+		done <- err
```

Comment:
How will this interact with normal pod shut down? won't we get in a weird loop where envoy is shutdown then restarted by pilot agent

Reply:
no. you can check `func (a *agent) Run(ctx context.Context) {`

when pod shutdown, the ctx is done, and the loop in it will break, and 
call this at last

```
		case <-ctx.Done():
			a.terminate()
			log.Info("Agent has successfully terminated")
			return
```

Reply:
:+1: I tested this, seems good